cache/lru_cache.cc

```diff
@@ -430,30 +430,81 @@ void LRUCacheShard::Promote(LRUHandle* e) {
 
   assert(secondary_handle->IsReady());
   e->SetIncomplete(false);
-  e->SetInCache(true);
+  e->SetInCache(false);
   e->value = secondary_handle->Value();
   e->CalcTotalCharge(secondary_handle->Size(), metadata_charge_policy_);
   delete secondary_handle;
 
-  // This call could fail if the cache is over capacity and
-  // strict_capacity_limit_ is true. In such a case, we don't want
-  // InsertItem() to free the handle, since the item is already in memory
-  // and the caller will most likely just read from disk if we erase it here.
   if (e->value) {
-    Cache::Handle* handle = reinterpret_cast<Cache::Handle*>(e);
-    Status s = InsertItem(e, &handle, /*free_handle_on_fail=*/false);
+    Status s;
+    // Insert a dummy handle and return a standalone handle to caller
+    // when secondary_cache_ is CompressedSecondaryCache and e is a
+    // standalone handle.
+    if (secondary_cache_ && secondary_cache_->SupportForceErase() &&
+        e->IsStandalone()) {
+      // Charge the standalone handle
+      autovector<LRUHandle*> last_reference_list;
+      {
+        DMutexLock l(mutex_);
+
+        // Free the space following strict LRU policy until enough space
+        // is freed or the lru list is empty.
+        EvictFromLRU(e->total_charge, &last_reference_list);
+
+        if ((usage_ + e->total_charge) > capacity_ && strict_capacity_limit_) {
+          e->Unref();
+          e->Free();
+          e = nullptr;
+        } else {
+          usage_ += e->total_charge;
+        }
+      }
+
+      // Try to insert the evicted entries into the secondary cache.
+      // Free the entries here outside of mutex for performance reasons.
+      for (auto entry : last_reference_list) {
+        if (secondary_cache_ && entry->IsSecondaryCacheCompatible() &&
+            !entry->IsInSecondaryCache()) {
+          secondary_cache_
+              ->Insert(entry->key(), entry->value, entry->info_.helper)
+              .PermitUncheckedError();
+        }
+        entry->Free();
+      }
```

Comment:
The whole for-loop must be copied from somewhere else. Can we define a helper function and call it once?